src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+        uint256 total = _add(amount, fees);
+        pool.withdraw(total);
+        daiJoin.join(address(this), total);
+
+        // normalized debt == erc20 DAI to pool (Vat rate for this ilk fixed to 1 RAY)
+
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(pool), address(pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(fees, RAY));
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        } else {
+            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, address(end_), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        }
+
+        emit Unwind(ilk, amount, fees);
+    }
+
+    /**
+        @notice Main function for updating a D3M position.
+        Determines the current state and either winds or unwinds as necessary.
+        @dev Winding the target position will be constrained by the Ilk debt
+        ceiling, the overall DSS debt ceiling and the maximum deposit by the
+        pool. Unwinding the target position will be constrained by the number
+        of assets available to be withdrawn from the pool.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function exec(bytes32 ilk) external {
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        pool.accrueIfNeeded();
+        uint256 availableAssets = pool.maxWithdraw();
+        uint256 currentAssets = pool.assetBalance();
+
+        if (vat.live() == 0) {
+            // MCD caged
+            require(end.debt() == 0, "DssDirectDepositHub/end-debt-already-set");
+            require(ilks[ilk].culled == 0, "DssDirectDepositHub/module-has-to-be-unculled-first");
+            _unwind(
+                ilk,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                Mode.MCD_CAGED,
+                currentAssets
+            );
+        } else if (ilks[ilk].tic != 0 || !pool.active() || !ilks[ilk].plan.active()) {
+            // pool caged
+            _unwind(
+                ilk,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                ilks[ilk].culled == 1
+                ? Mode.MODULE_CULLED
+                : Mode.NORMAL,
+                currentAssets
+            );
+        } else {
+            // Determine if it needs to unwind due to debt ceilings
+            (uint256 Art,,, uint256 line,) = vat.ilks(ilk);
```

Comment:
Move this to the top of the function and get rate here.

Add a require that `rate == 1 * RAY`

Reply:
3de376c